File: Array/345-reverse-vowels-of-a-string/reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U')) ;


        StringBuilder out = new StringBuilder(s);
        int r = out.length()-1;
        int l = 0;

        while(l<r) {
            char leftChar = out.charAt(l);
            char rightChar = out.charAt(r);
            if(!vowels.contains(leftChar)) {
                l++;
            } else if(!vowels.contains(rightChar)) {
                r--;
            } else {
                out.setCharAt(l,rightChar);
                out.setCharAt(r, leftChar);
                l++;
                r--;
            }
           
        }

        return out.toString();
    }
}

File: Array/345-reverse-vowels-of-a-string/reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set(['A', 'E', 'I', 'O', 'U', 'a', 'e', 'i', 'o', 'u'])
        arr = list(s)

        l = 0
        r = len(s)-1
        while l<r:
            if arr[r] in vowels and arr[l] in vowels:
                arr[r], arr[l] = arr[l], arr[r]
                r-=1
                l+=1
            elif arr[r] in vowels:
                l+=1
            elif arr[l] in vowels:
                r-=1
            else:
                l+=1
                r-=1
        return "".join(arr)